// deeplearn/dist/math/backends/tape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ndarray_1 = require("../ndarray");
var tape_util = require("./tape_util");
var Tape = (function () {
    function Tape(backend) {
        this.backend = backend;
        this.evaluatedTapeNodes = [];
        this.outputNodeMap = {};
    }
    Tape.prototype.addEvaluatedKernelNode = function (node) {
        this.outputNodeMap[node.output.id] = node;
        this.evaluatedTapeNodes.push(node);
    };
    Tape.prototype.gradientWrt = function (y, xs) {
        if (this.outputNodeMap[y.id] == null) {
            throw new Error("Cannot compute gradient: y is not part of this tape.");
        }
        var filteredNodes = tape_util.getFilteredNodesXToY(this.evaluatedTapeNodes, xs, y);
        var arrayAccumulatedGradientMap = {};
        arrayAccumulatedGradientMap[y.id] = ndarray_1.Scalar.new(1);
        tape_util.backpropagateGradients(this.backend, arrayAccumulatedGradientMap, filteredNodes);
        var gradients = [];
        for (var i = 0; i < xs.length; i++) {
            gradients.push(arrayAccumulatedGradientMap[xs[i].id]);
        }
        return gradients;
    };
    return Tape;
}());
exports.Tape = Tape;
